src/aiida_qe_xspec/workflows/xps.py

```diff
@@ -430,39 +419,26 @@ def get_builder_from_protocol(  # noqa
         else:
             builder.calc_binding_energy = orm.Bool(False)
         builder.clean_workdir = orm.Bool(inputs['clean_workdir'])
-        core_hole_pseudos = {}
-        gipaw_pseudos = {}
-        if elements_list:
-            elements_not_present = []
-            elements_present = [kind.symbol for kind in structure.kinds]
-            for element in elements_list:
-                if element not in elements_present:
-                    elements_not_present.append(element)
-
-            if len(elements_not_present) > 0:
-                raise ValueError(
-                    f'The following elements: {elements_not_present} are not present in the'
-                    f' structure ({elements_present}) provided.'
-                )
-            else:
-                builder.elements_list = orm.List(elements_list)
-                for element in elements_list:
-                    core_hole_pseudos[element] = pseudos[element]['core_hole']
-                    gipaw_pseudos[element] = pseudos[element]['gipaw']
-        elif atoms_list:
-            builder.atoms_list = orm.List(atoms_list)
-            for index in atoms_list:
-                element = structure.sites[index].kind_name
-                core_hole_pseudos[element] = pseudos[element]['core_hole']
-                gipaw_pseudos[element] = pseudos[element]['gipaw']
-        # if no elements list is given, we instead initalise the pseudos dict with all
-        # elements in the structure
-        else:
-            for element in pseudos:
-                core_hole_pseudos[element] = pseudos[element]['core_hole']
-                gipaw_pseudos[element] = pseudos[element]['gipaw']
+        elements_not_present = []
+        elements_present = [kind.symbol for kind in structure.kinds]
+        if core_levels is None:
+            core_levels = {}
+            for element, pseudos in core_hole_pseudos.items():
+                if element in structure.get_symbols_set():
+                    core_levels[element] = [key for key in pseudos.keys() if key != 'gipaw']
+        for label in core_levels:
+            element = label.split('_')[0]
+            if element not in elements_present:
+                elements_not_present.append(element)
```

Comment:
Since there is no input validation, `get_builder_from_protocol` will just pass in the keys of `core_hole_pseudos` for that element, regardless of whether or not those correspond to valid option, which can cause the program to except when trying to compile the final spectrum.

Some sort of input validation will be needed, in addition to deciding what inputs should be accepted as valid.